samza-core/src/main/java/org/apache/samza/operators/impl/PartitionByOperatorImpl.java

```diff
@@ -77,7 +77,8 @@ protected void handleInit(Config config, TaskContext context) {
       TaskCoordinator coordinator) {
     K key = keyFunction.apply(message);
     V value = valueFunction.apply(message);
-    collector.send(new OutgoingMessageEnvelope(systemStream, null, key, value));
+    Long partitionKey = key == null ? 0L : null;
```

Comment:
after checking with Xinyu, Kafka has a tricky behavior to do round robin  when both the partition key and key are null. It might be better to add a line of comment in code as well for future maintenance reference.

Reply:
This behavior in Kafka is pretty reasonable. A `null` key actually means that a key is absent (as opposed to a key being null) and the message is non-keyed. Imagine all non-keyed messages going to the same partition 0? That would be very sub-optimal. For stream-processing use-cases, we want to use `null` for `group`ing by key. Hence, we need this to get around this.